qos-service-impl/src/main/java/com/palantir/atlasdb/qos/QosResource.java

```diff
@@ -16,20 +16,77 @@
+import com.palantir.remoting3.jaxrs.JaxRsClient;
 
 public class QosResource implements QosService {
 
+    private final Optional<CassandraMetricsService> cassandraMetricClient;
     private Supplier<QosServiceRuntimeConfig> config;
 
     public QosResource(Supplier<QosServiceRuntimeConfig> config) {
         this.config = config;
+        this.cassandraMetricClient = config.get().qosCassandraMetricsConfig()
+                .map(metricsConfig -> Optional.of(JaxRsClient.create(
+                        CassandraMetricsService.class,
+                        "qos-service",
+                        ClientConfigurations.of(metricsConfig.cassandraServiceConfig()))))
+                .orElse(Optional.empty());
     }
 
     @Override
-    public long getLimit(String client) {
-        return config.get().clientLimits().getOrDefault(client, Long.MAX_VALUE);
+    public int readLimit(String client) {
+        QosClientLimitsConfig qosClientLimitsConfig = config.get().clientLimits().getOrDefault(client,
+                ImmutableQosClientLimitsConfig.builder().build());
+        return Ints.saturatedCast((long) getClientLimitMultiplier(qosClientLimitsConfig.clientPriority())
+                * qosClientLimitsConfig.limits().readBytesPerSecond());
+    }
+
+    @Override
+    public int writeLimit(String client) {
+        QosClientLimitsConfig qosClientLimitsConfig = config.get().clientLimits().getOrDefault(client,
+                ImmutableQosClientLimitsConfig.builder().build());
+        return Ints.saturatedCast((long) getClientLimitMultiplier(qosClientLimitsConfig.clientPriority())
+                * qosClientLimitsConfig.limits().writeBytesPerSecond());
+    }
+
+    private double getClientLimitMultiplier(QosPriority qosPriority) {
+        Optional<QosCassandraMetricsConfig> qosCassandraMetricsConfig = config.get().qosCassandraMetricsConfig();
+        if (qosCassandraMetricsConfig.isPresent() && cassandraMetricClient.isPresent()) {
+            QosCassandraMetricsConfig metricsConfig = qosCassandraMetricsConfig.get();
+
+            List<CassandraHealthMetricMeasurement> cassandraHealthMetricMeasurements =
+                    metricsConfig.cassandraHealthMetrics().stream().map(metric ->
+                            ImmutableCassandraHealthMetricMeasurement.builder()
+                                    .currentValue(cassandraMetricClient.get().getMetric(
+                                            metric.type(),
+                                            metric.name(),
+                                            metric.attribute(),
+                                            metric.additionalParams()))
+                                    .lowerLimit(metric.lowerLimit())
+                                    .upperLimit(metric.upperLimit())
+                                    .build())
+                            .collect(Collectors.toList());
+
+            return metricsConfig
+                    .throttlingStrategy()
```

Comment:
feels odd to have a stateful object coming out of the config. maybe the thing in the config should just be a factory?